src/services/textChanges.ts

```diff
@@ -0,0 +1,475 @@
+        printer.writeNode(EmitHint.Unspecified, node, sourceFile, writer);
+        return { text: writer.getText(), node: assignPositionsToNode(node) };
+    }
+
+    export function applyFormatting(nonFormattedText: NonFormattedText, sourceFile: SourceFile, initialIndentation: number, delta: number, rulesProvider: formatting.RulesProvider) {
+        const lineMap = computeLineStarts(nonFormattedText.text);
+        const file: SourceFileLike = {
+            text: nonFormattedText.text,
+            lineMap,
+            getLineAndCharacterOfPosition: pos => computeLineAndCharacterOfPosition(lineMap, pos)
+        };
+        const changes = formatting.formatNode(nonFormattedText.node, file, sourceFile.languageVariant, initialIndentation, delta, rulesProvider);
+        return applyChanges(nonFormattedText.text, changes);
+    }
+
+    export function applyChanges(text: string, changes: TextChange[]): string {
+        for (let i = changes.length - 1; i >= 0; i--) {
+            const change = changes[i];
+            text = `${text.substring(0, change.span.start)}${change.newText}${text.substring(textSpanEnd(change.span))}`;
+        }
+        return text;
+    }
+
+    function isTrivia(s: string) {
+        return skipTrivia(s, 0) === s.length;
+    }
+
+    const nullTransformationContext: TransformationContext = {
+        enableEmitNotification: noop,
+        enableSubstitution: noop,
+        endLexicalEnvironment: () => undefined,
+        getCompilerOptions: notImplemented,
+        getEmitHost: notImplemented,
+        getEmitResolver: notImplemented,
+        hoistFunctionDeclaration: noop,
+        hoistVariableDeclaration: noop,
+        isEmitNotificationEnabled: notImplemented,
+        isSubstitutionEnabled: notImplemented,
+        onEmitNode: noop,
+        onSubstituteNode: notImplemented,
+        readEmitHelpers: notImplemented,
+        requestEmitHelper: noop,
+        resumeLexicalEnvironment: noop,
+        startLexicalEnvironment: noop,
+        suspendLexicalEnvironment: noop
+    };
+
+    function assignPositionsToNode(node: Node): Node {
+        const visited = visitEachChild(node, assignPositionsToNode, nullTransformationContext, assignPositionsToNodeArray);
+        // create proxy node for non synthesized nodes
+        const newNode = nodeIsSynthesized(visited)
+            ? visited
+            : (Proxy.prototype = visited, new (<any>Proxy)());
+        newNode.pos = getPos(node);
+        newNode.end = getEnd(node);
+        return newNode;
+
+        function Proxy() { }
+    }
+
+    function assignPositionsToNodeArray(nodes: NodeArray<any>, visitor: Visitor, test?: (node: Node) => boolean, start?: number, count?: number) {
+        const visited = visitNodes(nodes, visitor, test, start, count);
+        if (!visited) {
+            return visited;
+        }
+        // clone nodearray if necessary
+        const nodeArray = visited === nodes ? createNodeArray(visited) : visited;

```

Comment:
`createNodeArray` returns the original nodes if the nodes are already `NodeArray`, which means the `getPos` and `getEnd` calls below can change the AST nodes that they shouldn't do.